starport/services/network/join.go

```diff
@@ -0,0 +1,167 @@
+package network
+
+import (
+	"context"
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	launchtypes "github.com/tendermint/spn/x/launch/types"
+	"github.com/tendermint/starport/starport/pkg/cosmosaddress"
+	"github.com/tendermint/starport/starport/pkg/cosmosutil"
+	"github.com/tendermint/starport/starport/pkg/events"
+)
+
+func (b *Builder) Join(
+	ctx context.Context,
+	chainHome string,
+	launchID uint64,
+	customGentx bool,
+	amount sdk.Coin,
+	peer string,
+	gentx []byte,
+	gentxInfo cosmosutil.Info) error {
+	if err := b.SendAccountRequest(ctx,
+		chainHome,
+		customGentx,
+		launchID,
+		amount); err != nil {
+		return err
+	}
+	return b.SendValidatorRequest(ctx, launchID, peer, gentx, gentxInfo)
+}
+
+// SendAccountRequest creates an add AddAccount request message
+func (b *Builder) SendAccountRequest(
+	ctx context.Context,
+	chainHome string,
+	isCustomGentx bool,
+	launchID uint64,
+	amount sdk.Coin,
+) error {
+	address := b.account.Address(SPNAddressPrefix)
+	spnAddress, err := cosmosaddress.ChangePrefix(address, SPNAddressPrefix)
+	if err != nil {
+		return err
+	}
+
+	b.ev.Send(events.New(events.StatusOngoing, "Verifying account already exists "+spnAddress))
+
+	// if is custom gentx path, avoid to check account into genesis from the home folder
+	accExist := false
+	if !isCustomGentx {
+		accExist, err = cosmosutil.CheckGenesisContainsAddress(chainHome, spnAddress)
+		if err != nil {
+			return err
+		}
+	}
+	// check if account exists as a genesis account in SPN chain launch information
+	if !accExist && !b.hasAccount(ctx, launchID, spnAddress) {
+		msg := launchtypes.NewMsgRequestAddAccount(
+			spnAddress,
+			launchID,
+			sdk.NewCoins(amount),
+		)
+
+		b.ev.Send(events.New(events.StatusOngoing, "Broadcasting account transactions"))
+		res, err := b.cosmos.BroadcastTx(b.account.Name, msg)
+		if err != nil {
+			return err
+		}
+
+		var requestRes launchtypes.MsgRequestAddAccountResponse
+		if err := res.Decode(&requestRes); err != nil {
+			return err
+		}
+		b.ev.Send(events.New(events.StatusDone, "MsgRequestAddAccount transactions sent"))
+
+		if requestRes.AutoApproved {
+			b.ev.Send(events.New(events.StatusDone, "Account added to the network by the coordinator!"))
+		} else {
+			b.ev.Send(events.New(events.StatusDone,
+				fmt.Sprintf("Request %d to add account to the network has been submitted!",
+					requestRes.RequestID),
+			))
+		}
+		return nil
+	}
+
+	b.ev.Send(events.New(events.StatusDone, "Account already exist"))
+	return err
+}
+
+// SendValidatorRequest creates the RequestAddValidator message into the SPN
+func (b *Builder) SendValidatorRequest(
```

Comment:
Builder.SendValidatorRequest to Blockchain.SendValidatorRequest and same for others?